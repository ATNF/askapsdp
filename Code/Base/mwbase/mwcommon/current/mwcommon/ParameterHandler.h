/// @file
/// @brief Handle a LOFAR .parset file
///
/// @copyright (c) 2007 CSIRO
/// Australia Telescope National Facility (ATNF)
/// Commonwealth Scientific and Industrial Research Organisation (CSIRO)
/// PO Box 76, Epping NSW 1710, Australia
/// atnf-enquiries@csiro.au
///
/// This file is part of the ASKAP software distribution.
///
/// The ASKAP software distribution is free software: you can redistribute it
/// and/or modify it under the terms of the GNU General Public License as
/// published by the Free Software Foundation; either version 2 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program; if not, write to the Free Software
/// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
///
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef ASKAP_MWCOMMON_PARAMETERHANDLER_H
#define ASKAP_MWCOMMON_PARAMETERHANDLER_H

#include <Common/ParameterSet.h>
#include <Blob/BlobIStream.h>
#include <Blob/BlobOStream.h>

namespace askap { namespace mwbase {

  /// @ingroup mwcontrol
  /// @brief Handle a LOFAR .parset file

  /// This class handles the processing of a LOFAR .parset file
  /// It augments the LOFAR ParameterSet class with functions that can deal
  /// with undefined parameters. There is a set of functions that return
  /// a default value if undefined and a set of functions that leave the
  /// value untouched if undefined.

  class ParameterHandler
  {
  public:
    explicit ParameterHandler (const LOFAR::ParameterSet&);

    //// Get a parameter value.
    //// An exception is thrown if it does not exist.
    //// @{
    std::string getString (const std::string& parm) const;
    double getDouble (const std::string& parm) const;
    unsigned getUint (const std::string& parm) const;
    bool getBool (const std::string& parm) const;
    std::vector<std::string> getStringVector (const std::string& parm) const;
    //// @}

    //// Get a parameter value.
    //// If it does not exist, the default value is used instead.
    //// @{
    std::string getString (const std::string& parm,
			   const std::string& defVal) const;
    double getDouble (const std::string& parm,
		      double defVal) const;
    unsigned getUint (const std::string& parm,
		      unsigned defVal) const;
    bool getBool (const std::string& parm,
		  bool defVal) const;
    std::vector<std::string> getStringVector
      (const std::string& parm, const std::vector<std::string>& defVal) const;
    //// @}

    //// Get a parameter value and fill \a value with it.
    //// If it does not exist, nothing is done.
    //// @{
    void fillString (const std::string& parm,
		     std::string& value) const;
    void fillDouble (const std::string& parm,
		     double& value) const;
    void fillUint (const std::string& parm,
		   unsigned& value) const;
    void fillBool (const std::string& parm,
		   bool& value) const;
    void fillStringVector (const std::string& parm,
			   std::vector<std::string>& value) const;
    //// @}

  protected:
    LOFAR::ParameterSet itsParms;
  };


  // Write/read a ParameterSet into/from a blob.
  // @{
  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream&,
                                 const LOFAR::ParameterSet&);
  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream&,
                                 LOFAR::ParameterSet&);
  // @}

  inline std::string ParameterHandler::getString (const std::string& parm) const
    { return itsParms.getString (parm); }
  inline double ParameterHandler::getDouble (const std::string& parm) const
    { return itsParms.getDouble (parm); }
  inline unsigned ParameterHandler::getUint (const std::string& parm) const
    { return itsParms.getUint32 (parm); }
  inline bool ParameterHandler::getBool (const std::string& parm) const
    { return itsParms.getBool (parm); }
  inline std::vector<std::string> ParameterHandler::getStringVector
  (const std::string& parm) const
    { return itsParms.getStringVector (parm); }

}} /// end namespaces

#endif
