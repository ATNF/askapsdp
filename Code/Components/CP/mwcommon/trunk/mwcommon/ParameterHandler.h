/// @file
/// @brief Handle a LOFAR .parset file
///
/// @copyright (c) 2007 CONRAD, All Rights Reserved.
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef CONRAD_MWCOMMON_PARAMETERHANDLER_H
#define CONRAD_MWCOMMON_PARAMETERHANDLER_H

#include <APS/ParameterSet.h>
#include <Blob/BlobIStream.h>
#include <Blob/BlobOStream.h>

namespace conrad { namespace cp {

  /// @ingroup mwcontrol
  /// @brief Handle a LOFAR .parset file

  /// This class handles the processing of a LOFAR .parset file
  /// It augments the LOFAR ParameterSet class with functions that can deal
  /// with undefined parameters. There is a set of functions that return
  /// a default value if undefined and a set of functions that leave the
  /// value untouched if undefined.

  class ParameterHandler
  {
  public:
    explicit ParameterHandler (const LOFAR::ACC::APS::ParameterSet&);

    //// Get a parameter value.
    //// An exception is thrown if it does not exist.
    //// @{
    std::string getString (const std::string& parm) const;
    double getDouble (const std::string& parm) const;
    unsigned getUint (const std::string& parm) const;
    bool getBool (const std::string& parm) const;
    std::vector<std::string> getStringVector (const std::string& parm) const;
    //// @}

    //// Get a parameter value.
    //// If it does not exist, the default value is used instead.
    //// @{
    std::string getString (const std::string& parm,
			   const std::string& defVal) const;
    double getDouble (const std::string& parm,
		      double defVal) const;
    unsigned getUint (const std::string& parm,
		      unsigned defVal) const;
    bool getBool (const std::string& parm,
		  bool defVal) const;
    std::vector<std::string> getStringVector
      (const std::string& parm, const std::vector<std::string>& defVal) const;
    //// @}

    //// Get a parameter value and fill \a value with it.
    //// If it does not exist, nothing is done.
    //// @{
    void fillString (const std::string& parm,
		     std::string& value) const;
    void fillDouble (const std::string& parm,
		     double& value) const;
    void fillUint (const std::string& parm,
		   unsigned& value) const;
    void fillBool (const std::string& parm,
		   bool& value) const;
    void fillStringVector (const std::string& parm,
			   std::vector<std::string>& value) const;
    //// @}

  protected:
    LOFAR::ACC::APS::ParameterSet itsParms;
  };


  // Write/read a ParameterSet into/from a blob.
  // @{
  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream&,
                                 const LOFAR::ACC::APS::ParameterSet&);
  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream&,
                                 LOFAR::ACC::APS::ParameterSet&);
  // @}

  inline std::string ParameterHandler::getString (const std::string& parm) const
    { return itsParms.getString (parm); }
  inline double ParameterHandler::getDouble (const std::string& parm) const
    { return itsParms.getDouble (parm); }
  inline unsigned ParameterHandler::getUint (const std::string& parm) const
    { return itsParms.getUint32 (parm); }
  inline bool ParameterHandler::getBool (const std::string& parm) const
    { return itsParms.getBool (parm); }
  inline std::vector<std::string> ParameterHandler::getStringVector
  (const std::string& parm) const
    { return itsParms.getStringVector (parm); }

}} /// end namespaces

#endif
