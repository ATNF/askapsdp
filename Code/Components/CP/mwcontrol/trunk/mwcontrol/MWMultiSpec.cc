//#  MWMultiSpec.cc: 
//#
//#  Copyright (C) 2007
//#
//#  $Id$

#include <mwcontrol/MWMultiSpec.h>
#include <mwcommon/MWError.h>
#include <APS/ParameterSet.h>

using namespace LOFAR;
using namespace LOFAR::ACC::APS;
using namespace std;

namespace conrad { namespace cp {

  MWMultiSpec::MWMultiSpec()
    : MWSpec()
  {}

  MWMultiSpec::MWMultiSpec(const string& name,
			   const ParameterSet& parset,
			   const MWSpec* parent)
    : MWSpec(name, parset, parent)
  {
    // This multispec consists of the following specs.
    vector<string> specs(parset.getStringVector("Step." + name + ".Steps"));
    // Create a new spec for each name in \a specs.
    for (uint i=0; i<specs.size(); ++i) {
      infiniteRecursionCheck (specs[i]);
      itsSpecs.push_back (MWSpec::create(specs[i], parset, this));
    }
  }


  MWMultiSpec::~MWMultiSpec()
  {}

  void MWMultiSpec::visit (MWSpecVisitor& visitor) const
  {
    visitor.visitMulti (*this);
  }

  void MWMultiSpec::print(ostream& os, const string& indent) const
  {
    printSpec (os, indent, "Multi");
    string indent2 = indent + ". ";
    for (const_iterator it=begin(); it!=end(); ++it) {
      os << endl;
      (*it)->print (os, indent2);
    }
  }

  void MWMultiSpec::infiniteRecursionCheck(const string& name) const
  {
    if (name == getName()) {
      CONRADTHROW (MWError, 
		   "Infinite recursion detected in definition of MWSpec \""
		   << name << "\". Please check your ParameterSet file.");
    }
    const MWMultiSpec* parent;
    if ((parent = dynamic_cast<const MWMultiSpec*>(getParent())) != 0) {
      parent->infiniteRecursionCheck(name);
    }
  }

}} // end namespaces
