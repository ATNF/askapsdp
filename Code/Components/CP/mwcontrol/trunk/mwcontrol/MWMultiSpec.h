//# MWMultiSpec.h: Derived composite class of the MWSpec composite pattern.
//#
//# Copyright (C) 2007
//#
//# $Id$

#ifndef CONRAD_MWCONTROL_MWMULTISPEC_H
#define CONRAD_MWCONTROL_MWMULTISPEC_H

/// \file
/// Derived composite class of the MWSpec composite pattern.

//# Includes
#include <mwcontrol/MWSpec.h>
#include <list>

namespace conrad { namespace cp {

  /// This is the so-called \e composite class in the composite pattern (see
  /// Gamma, 1995). The composite class contains pointers to zero or more
  /// MWSpec (component) objects.

  class MWMultiSpec : public MWSpec
  {
  public:
    /// Construct a MWMultiSpec. \a name identifies the step name in the
    /// parameter set file. It does \e not uniquely identify the step \e
    /// object being created. The third argument is used to pass a
    /// backreference to the parent MWSpec object.
    MWMultiSpec(const std::string& name,
		const LOFAR::ACC::APS::ParameterSet& parset,
		const MWSpec* parent);

    /// Default constructor. Construct an empty MWMultiSpec object and make
    /// it a child of the MWSpec object \a parent.
    explicit MWMultiSpec(const MWSpec* parent = 0) : MWSpec(parent) {}

    virtual ~MWMultiSpec();

    /// Add a step.
    void push_back (const MWSpec::ShPtr& spec)
      { itsSpecs.push_back (spec); }

    /// Visit the object, so the visitor can process it.
    virtual void visit (MWSpecVisitor&) const;

    /// Print the contents of \c *this in human readable form into the output
    /// stream \a os.
    virtual void print (std::ostream& os, const std::string& indent) const;

    /// Define functions and so to iterate.
    /// @{
    typedef std::list<MWSpec::ShPtr>::const_iterator const_iterator;
    const_iterator begin() const
      { return itsSpecs.begin(); }
    const_iterator end() const
      { return itsSpecs.end(); }
    /// @}

  private:
    /// Check to see if there's an infinite recursion present in the
    /// definition of a MWMultiSpec. This can happen when one of the specs
    /// (identified by the argument \a name) defining the MWMultiSpec refers
    /// directly or indirectly to that same MWMultiSpec. 
    void infiniteRecursionCheck (const std::string& name) const;

    /// Vector holding a sequence of MWSpecs.
    std::list<MWSpec::ShPtr> itsSpecs;
  };

  /// @}
    
}} /// end namespaces

#endif
