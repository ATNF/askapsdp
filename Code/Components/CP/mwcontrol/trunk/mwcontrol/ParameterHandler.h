/// @file
/// @brief Handle a LOFAR .parset file
///
/// @copyright (c) 2007 CONRAD, All Rights Reserved.
/// @author Ger van Diepen (diepen AT astron nl)
///
//# $Id$

#ifndef CONRAD_MWCONTROL_PARAMETERHANDLER_H
#define CONRAD_MWCONTROL_PARAMETERHANDLER_H

#include <APS/ParameterSet.h>
#include <Blob/BlobIStream.h>
#include <Blob/BlobOStream.h>

namespace conrad { namespace cp {

  /// Forward Declarations
  class MWMultiSpec;
  class MWStrategySpec;

  /// @ingroup mwcontrol
  /// @brief Handle a LOFAR .parset file

  /// This class handles the processing of a LOFAR .parset file
  /// It has two sets of functions:
  /// <ul>
  ///  <li> It augments the LOFAR ParameterSet class with functions that can deal
  ///       with undefined parameters. There is a set of functions that return
  ///       a default value if undefined and a set of functions that leave the
  ///       value untouched if undefined.
  ///  <li> Functions to retrieve specific info from the .parset file.
  ///       These can be initial info, strategy specifications (MWStrategySpec),
  ///       and step specifications (MWStep).
  /// </ul>

  class ParameterHandler
  {
  public:
    explicit ParameterHandler (const LOFAR::ACC::APS::ParameterSet&);

    //// Get the number of VDS parts.
    int getNParts() const;

    //// Get the initial parameters.
    void getInitInfo (std::string& msName,
		      std::string& inputColumn,
		      std::string& skyParameterDB,
		      std::string& instrumentParameterDB,
		      unsigned& subBand,
		      bool& calcUVW) const;

    //// Get all strategies specifications from the parameters.
    std::vector<MWStrategySpec> getStrategies() const;

    //// Get all step specifications of a strategy from the parameters.
    MWMultiSpec getSteps (const std::string& strategyName) const;

    //// Get a parameter value.
    //// An exception is thrown if it does not exist.
    //// @{
    std::string getString (const std::string& parm) const;
    double getDouble (const std::string& parm) const;
    unsigned getUint (const std::string& parm) const;
    bool getBool (const std::string& parm) const;
    std::vector<std::string> getStringVector (const std::string& parm) const;
    //// @}

    //// Get a parameter value.
    //// If it does not exist, the default value is used instead.
    //// @{
    std::string getString (const std::string& parm,
			   const std::string& defVal) const;
    double getDouble (const std::string& parm,
		      double defVal) const;
    unsigned getUint (const std::string& parm,
		      unsigned defVal) const;
    bool getBool (const std::string& parm,
		  bool defVal) const;
    std::vector<std::string> getStringVector
      (const std::string& parm, const std::vector<std::string>& defVal) const;
    //// @}

    //// Get a parameter value and fill \a value with it.
    //// If it does not exist, nothing is done.
    //// @{
    void fillString (const std::string& parm,
		     std::string& value) const;
    void fillDouble (const std::string& parm,
		     double& value) const;
    void fillUint (const std::string& parm,
		   unsigned& value) const;
    void fillBool (const std::string& parm,
		   bool& value) const;
    void fillStringVector (const std::string& parm,
			   std::vector<std::string>& value) const;
    //// @}

  private:
    LOFAR::ACC::APS::ParameterSet itsParms;
  };


  // Write/read a ParameterSet into/from a blob.
  // @{
  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream&,
                                 const LOFAR::ACC::APS::ParameterSet&);
  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream&,
                                 LOFAR::ACC::APS::ParameterSet&);
  // @}

  inline std::string ParameterHandler::getString (const std::string& parm) const
    { return itsParms.getString (parm); }
  inline double ParameterHandler::getDouble (const std::string& parm) const
    { return itsParms.getDouble (parm); }
  inline unsigned ParameterHandler::getUint (const std::string& parm) const
    { return itsParms.getUint32 (parm); }
  inline bool ParameterHandler::getBool (const std::string& parm) const
    { return itsParms.getBool (parm); }
  inline std::vector<std::string> ParameterHandler::getStringVector
  (const std::string& parm) const
    { return itsParms.getStringVector (parm); }

}} /// end namespaces

#endif
