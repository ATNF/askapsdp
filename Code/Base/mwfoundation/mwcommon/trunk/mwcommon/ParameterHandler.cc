//#  ParameterHandler.cc: 
//#
/// @copyright (c) 2007 CSIRO
/// Australia Telescope National Facility (ATNF)
/// Commonwealth Scientific and Industrial Research Organisation (CSIRO)
/// PO Box 76, Epping NSW 1710, Australia
/// atnf-enquiries@csiro.au
///
/// This file is part of the ASKAP software distribution.
///
/// The ASKAP software distribution is free software: you can redistribute it
/// and/or modify it under the terms of the GNU General Public License as
/// published by the Free Software Foundation; either version 2 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program; if not, write to the Free Software
/// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
///
//#
//#  $Id$

#include <mwcommon/ParameterHandler.h>

using namespace LOFAR::ACC::APS;
using namespace std;

namespace askap { namespace mwbase {

  ParameterHandler::ParameterHandler (const ParameterSet& parSet)
    : itsParms (parSet)
  {}

  string ParameterHandler::getString (const string& parm,
				      const string& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getString (parm);
    }
    return defVal;
  }

  double ParameterHandler::getDouble (const string& parm,
				      double defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getDouble (parm);
    }
    return defVal;
  }

  unsigned ParameterHandler::getUint (const string& parm,
				      unsigned defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getUint32 (parm);
    }
    return defVal;
  }

  bool ParameterHandler::getBool (const string& parm,
				  bool defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getBool (parm);
    }
    return defVal;
  }

  vector<string> ParameterHandler::getStringVector
  (const string& parm, const vector<string>& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getStringVector (parm);
    }
    return defVal;
  }

  void ParameterHandler::fillString (const string& parm,
				     string& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getString (parm);
    }
  }

  void ParameterHandler::fillDouble (const string& parm,
				     double& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getDouble (parm);
    }
  }

  void ParameterHandler::fillUint (const string& parm,
				   unsigned& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getUint32 (parm);
    }
  }

  void ParameterHandler::fillBool (const string& parm,
				   bool& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getBool (parm);
    }
  }

  void ParameterHandler::fillStringVector (const string& parm,
					   vector<string>& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getStringVector (parm);
    }
  }


  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream& bs,
                                 const LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.putStart ("ParameterSet", 1);
    bs << static_cast<LOFAR::uint32>(m.size());
    for (LOFAR::ACC::APS::ParameterSet::const_iterator it=m.begin();
         it!=m.end();
         ++it) {
      bs << it->first << it->second;
    }
    bs.putEnd();
    return bs;
  }

  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream& bs,
                                 LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.getStart ("ParameterSet");
    m.clear();
    LOFAR::uint32 size;
    bs >> size;
    std::string k,v;
    for (LOFAR::uint32 i=0; i<size; ++i) {
      bs >> k >> v;
      m.add (k, v);
    }
    bs.getEnd();
    return bs;
  }

}} // end namespaces
