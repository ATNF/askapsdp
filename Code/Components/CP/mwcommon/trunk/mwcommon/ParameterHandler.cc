//#  ParameterHandler.cc: 
//#
//#  Copyright (C) 2007
//#
//#  $Id$

#include <mwcommon/ParameterHandler.h>

using namespace LOFAR::ACC::APS;
using namespace std;

namespace conrad { namespace cp {

  ParameterHandler::ParameterHandler (const ParameterSet& parSet)
    : itsParms (parSet)
  {}

  string ParameterHandler::getString (const string& parm,
				      const string& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getString (parm);
    }
    return defVal;
  }

  double ParameterHandler::getDouble (const string& parm,
				      double defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getDouble (parm);
    }
    return defVal;
  }

  unsigned ParameterHandler::getUint (const string& parm,
				      unsigned defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getUint32 (parm);
    }
    return defVal;
  }

  bool ParameterHandler::getBool (const string& parm,
				  bool defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getBool (parm);
    }
    return defVal;
  }

  vector<string> ParameterHandler::getStringVector
  (const string& parm, const vector<string>& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getStringVector (parm);
    }
    return defVal;
  }

  void ParameterHandler::fillString (const string& parm,
				     string& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getString (parm);
    }
  }

  void ParameterHandler::fillDouble (const string& parm,
				     double& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getDouble (parm);
    }
  }

  void ParameterHandler::fillUint (const string& parm,
				   unsigned& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getUint32 (parm);
    }
  }

  void ParameterHandler::fillBool (const string& parm,
				   bool& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getBool (parm);
    }
  }

  void ParameterHandler::fillStringVector (const string& parm,
					   vector<string>& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getStringVector (parm);
    }
  }


  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream& bs,
                                 const LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.putStart ("ParameterSet", 1);
    bs << static_cast<LOFAR::uint32>(m.size());
    for (LOFAR::ACC::APS::ParameterSet::const_iterator it=m.begin();
         it!=m.end();
         ++it) {
      bs << it->first << it->second;
    }
    bs.putEnd();
    return bs;
  }

  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream& bs,
                                 LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.getStart ("ParameterSet");
    m.clear();
    LOFAR::uint32 size;
    bs >> size;
    std::string k,v;
    for (LOFAR::uint32 i=0; i<size; ++i) {
      bs >> k >> v;
      m.add (k, v);
    }
    bs.getEnd();
    return bs;
  }

}} // end namespaces
