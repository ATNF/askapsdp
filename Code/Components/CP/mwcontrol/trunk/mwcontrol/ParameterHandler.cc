//#  ParameterHandler.cc: 
//#
//#  Copyright (C) 2007
//#
//#  $Id$

#include <mwcontrol/ParameterHandler.h>
#include <mwcontrol/MWStrategySpec.h>
#include <mwcontrol/MWMultiSpec.h>

using namespace LOFAR::ACC::APS;
using namespace std;

namespace conrad { namespace cp {

  ParameterHandler::ParameterHandler (const ParameterSet& parSet)
    : itsParms (parSet)
  {}

  void ParameterHandler::getInitInfo (string& msName,
				      string& inputColumn,
				      string& skyParameterDB,
				      string& instrumentParameterDB,
				      unsigned& subBand,
				      bool& calcUVW) const
  {
    msName                = getString ("DataSet");
    inputColumn           = getString ("Strategy.InputData", "DATA");
    skyParameterDB        = getString ("ParmDB.LocalSky");
    instrumentParameterDB = getString ("ParmDB.Instrument");
    subBand               = getUint   ("SubBandID", 0);
    calcUVW               = getBool   ("CalcUVW", false);
  }

  int ParameterHandler::getNParts() const
  {
    return getUint ("NNode", 1);
  }

  vector<MWStrategySpec> ParameterHandler::getStrategies() const
  {
    // Get all strategy names.
    // Default is 'Strategy'.
    vector<string> defName(1, "Strategy");
    vector<string> strategyNames(getStringVector("Strategies", defName));
    // Create a new strategy specification object for each name.
    vector<MWStrategySpec> specs;
    for (unsigned i=0; i<strategyNames.size(); ++i) {
      specs.push_back (MWStrategySpec (strategyNames[i], itsParms));
    }
    return specs;
  }

  MWMultiSpec ParameterHandler::getSteps (const string& name) const
  {
    // Get all step names.
    vector<string> stepNames(itsParms.getStringVector(name+".Steps"));
    // Create a new step specification object for each name.
    MWMultiSpec specs;
    for (unsigned i=0; i<stepNames.size(); ++i) {
      specs.push_back (MWSpec::create (stepNames[i], itsParms, 0));
    }
    return specs;
  }

  string ParameterHandler::getString (const string& parm,
				      const string& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getString (parm);
    }
    return defVal;
  }

  double ParameterHandler::getDouble (const string& parm,
				      double defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getDouble (parm);
    }
    return defVal;
  }

  unsigned ParameterHandler::getUint (const string& parm,
				      unsigned defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getUint32 (parm);
    }
    return defVal;
  }

  bool ParameterHandler::getBool (const string& parm,
				  bool defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getBool (parm);
    }
    return defVal;
  }

  vector<string> ParameterHandler::getStringVector
  (const string& parm, const vector<string>& defVal) const
  {
    if (itsParms.isDefined(parm)) {
      return itsParms.getStringVector (parm);
    }
    return defVal;
  }

  void ParameterHandler::fillString (const string& parm,
				     string& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getString (parm);
    }
  }

  void ParameterHandler::fillDouble (const string& parm,
				     double& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getDouble (parm);
    }
  }

  void ParameterHandler::fillUint (const string& parm,
				   unsigned& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getUint32 (parm);
    }
  }

  void ParameterHandler::fillBool (const string& parm,
				   bool& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getBool (parm);
    }
  }

  void ParameterHandler::fillStringVector (const string& parm,
					   vector<string>& value) const
  {
    if (itsParms.isDefined(parm)) {
      value = itsParms.getStringVector (parm);
    }
  }


  LOFAR::BlobOStream operator<< (LOFAR::BlobOStream& bs,
                                 const LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.putStart ("ParameterSet", 1);
    bs << static_cast<LOFAR::uint32>(m.size());
    for (LOFAR::ACC::APS::ParameterSet::const_iterator it=m.begin();
         it!=m.end();
         ++it) {
      bs << it->first << it->second;
    }
    bs.putEnd();
    return bs;
  }

  LOFAR::BlobIStream operator>> (LOFAR::BlobIStream& bs,
                                 LOFAR::ACC::APS::ParameterSet& m)
  {
    bs.getStart ("ParameterSet");
    m.clear();
    LOFAR::uint32 size;
    bs >> size;
    std::string k,v;
    for (LOFAR::uint32 i=0; i<size; ++i) {
      bs >> k >> v;
      m.add (k, v);
    }
    bs.getEnd();
    return bs;
  }

}} // end namespaces
