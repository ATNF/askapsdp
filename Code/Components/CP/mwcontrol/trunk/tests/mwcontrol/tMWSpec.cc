//#  tMWStep.cc: test program for the MWStep class
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tMWStep.cc,v 1.6 2006/10/03 15:54:07 loose Exp $

#include <mwcontrol/MWMultiSpec.h>
#include <mwcontrol/ParameterHandler.h>

using namespace LOFAR::ACC::APS;
using namespace conrad::cp;
using namespace std;

int main()
{
  try {
    ParameterHandler psh (ParameterSet("tMWSpec.in"));
    psh.getSteps("Strategy").print (cout, "");
    cout << endl;
  } catch (LOFAR::Exception& e) {
    cout << "Unexpected: " << e.what() << endl;
    return 1;
  }
  cout << "OK" << endl;
  return 0;
}
