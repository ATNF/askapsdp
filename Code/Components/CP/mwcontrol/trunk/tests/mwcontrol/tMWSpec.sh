#!/bin/sh
$lofar_sharedir/runtest.sh tMWSpec > tMWSpec.log 2>&1
