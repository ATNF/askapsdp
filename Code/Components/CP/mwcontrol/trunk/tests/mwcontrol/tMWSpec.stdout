Multi specification: 
 Full name: 
 Station1:         []
 Station2:         []
 Corr selection:   
 Correlation type: []
 Integration:      1e+30 Hz,  1e+30 s
 Sources:          []
 Extra sources:    []
 Instrument model: []
. Multi specification: xyz1
.  Full name: xyz1
.  Station1:         [CS*, CS*, CS*, RS1*, RS1*, RS2*]
.  Station2:         [CS*, RS1*, RS2*, RS1*, RS2*, RS2*]
.  Corr selection:   CROSS
.  Correlation type: [XX, XY, YX, YY]
.  Integration:      2 Hz,  0.5 s
.  Sources:          [3C343]
.  Extra sources:    [M81]
.  Instrument model: [Bandpass, DirGain, Phase]
. . Solve specification: sl1
. .  Full name: xyz1.sl1
. .  Station1:         [CS*, RS1*]
. .  Station2:         [RS1*, RS2*]
. .  Corr selection:   CROSS
. .  Correlation type: [XX, XY, YX, YY]
. .  Integration:      2 Hz,  0.5 s
. .  Sources:          [3C343]
. .  Extra sources:    [M81]
. .  Instrument model: [Bandpass, Phase]
. .  Solve: 
. .   Max nr. of iterations:  10
. .   Convergence threshold:  1e-07
. .   Min fraction converged: 0.95
. .   Solvable parameters:    [PHASE:*]
. .   Excluded parameters:    []
. .   Domain size :           1000 Hz,  1 s
. . Correct specification: sl2
. .  Full name: xyz1.sl2
. .  Station1:         [CS*, CS*, CS*, RS1*, RS1*, RS2*]
. .  Station2:         [CS*, RS1*, RS2*, RS1*, RS2*, RS2*]
. .  Corr selection:   CROSS
. .  Correlation type: [XX, XY, YX, YY]
. .  Integration:      2 Hz,  0.5 s
. .  Sources:          [3C343]
. .  Extra sources:    [M81]
. .  Instrument model: [DirGain, Phase]
. .  Output data: OUTDATA2
. Multi specification: xyz2
.  Full name: xyz2
.  Station1:         [CS*, CS*, RS1*]
.  Station2:         [CS*, RS1*, RS1*]
.  Corr selection:   
.  Correlation type: []
.  Integration:      1e+30 Hz,  1e+30 s
.  Sources:          [3C343]
.  Extra sources:    []
.  Instrument model: [Bandpass]
. . Solve specification: sl1
. .  Full name: xyz2.sl1
. .  Station1:         [CS*, RS1*]
. .  Station2:         [RS1*, RS2*]
. .  Corr selection:   
. .  Correlation type: []
. .  Integration:      1e+30 Hz,  1e+30 s
. .  Sources:          [3C343]
. .  Extra sources:    []
. .  Instrument model: [Bandpass, Phase]
. .  Solve: 
. .   Max nr. of iterations:  10
. .   Convergence threshold:  1e-07
. .   Min fraction converged: 0.95
. .   Solvable parameters:    [PHASE:*]
. .   Excluded parameters:    []
. .   Domain size :           1000 Hz,  1 s
. . Predict specification: sl3
. .  Full name: xyz2.sl3
. .  Station1:         [CS*, CS*, RS1*]
. .  Station2:         [CS*, RS1*, RS1*]
. .  Corr selection:   
. .  Correlation type: []
. .  Integration:      1e+30 Hz,  1e+30 s
. .  Sources:          [3C343]
. .  Extra sources:    []
. .  Instrument model: [Phase]
. .  Output data: OUTDATA3
OK
